astropy/io/registry/core.py

```diff
@@ -2,23 +2,28 @@
 
 import os
 import sys
+from typing import TypeVar
 
 from .base import IORegistryError, _UnifiedIORegistryBase
 
 __all__ = ["UnifiedIORegistry", "UnifiedInputRegistry", "UnifiedOutputRegistry"]
 
 
 PATH_TYPES = (str, os.PathLike)  # TODO! include bytes
+T = TypeVar("T", bound=str)
 
 
-def _expand_user_in_args(args):
+def _expand_user_in_arg0(arg0: T | os.PathLike[T]) -> T:
     # Conservatively attempt to apply `os.path.expanduser` to the first
     # argument, which can be either a path or the contents of a table.
-    if len(args) and isinstance(args[0], PATH_TYPES):
-        ex_user = os.path.expanduser(args[0])
-        if ex_user != args[0] and os.path.exists(os.path.dirname(ex_user)):
-            args = (ex_user,) + args[1:]
-    return args
+    if (
+        isinstance(arg0, PATH_TYPES)
+        and (ex_user := os.path.expanduser(arg0)) != arg0
+        and os.path.exists(os.path.dirname(ex_user))
+    ):
+        return ex_user
+    else:
+        return arg0
```

Comment:
So this function might still return a pathlib object or str/bytes based on some hard to predict logic ?

Reply:
hum, yes, I suppose I missed that. However, the condition isn't so hard to predict: it happens if and only if the input is a `Path` whose parent directory doesn't exist... I'll fix it.